metricproviders/kayenta/kayenta.go

```
@@ -0,0 +1,257 @@
+
+	configIdLookupURLFormat = `%s/canaryConfig?application=%s&configurationAccountName=%s`
+
+	jobURLFormat = `%s/canary/%s?application=%s&metricsAccountName=%s&configurationAccountName=%s&storageAccountName=%s`
+
+	jobPayloadFormat = `
+							{
+								"scopes": {
+										%s
+								},
+                                "thresholds" : {
+                                    "pass": %d,
+                                    "marginal": %d
+                                }
+                            }`
+
+	resumeDelay          time.Duration = 15 * time.Second
+	httpConnectionTimout time.Duration = 15 * time.Second
+	scopeFormat                        = `"%s":{"controlScope": %s, "experimentScope": %s}`
+)
+
+type Provider struct {
+	logCtx log.Entry
+	client http.Client
+}
+
+type canaryConfig struct {
+	Id                  string
+	Name                string
+	UpdatedTimestamp    int
+	UpdatedTimestampIso string
+	Applications        []string
+}
+
+// Type incidates provider is a kayenta provider
+func (p *Provider) Type() string {
+	return ProviderType
+}
+
+func getCanaryConfigId(metric v1alpha1.Metric, p *Provider) (string, error) {
+	canaryConfigId := ""
+	configIdLookupURL := fmt.Sprintf(configIdLookupURLFormat, metric.Provider.Kayenta.Address, metric.Provider.Kayenta.Application, metric.Provider.Kayenta.StorageAccountName)
+
+	response, err := p.client.Get(configIdLookupURL)
+	if err != nil || response.Body == nil || response.StatusCode != 200 {
+		if err == nil {
+			err = errors.New("Invalid Response")
+		}
+		return canaryConfigId, err
+	}
+
+	data, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return "", err
+	}
+
+	var cc []canaryConfig
+
+	err = json.Unmarshal(data, &cc)
+	if err != nil {
+		return canaryConfigId, err
+	}
+
+	for _, s := range cc {
+		if s.Name == metric.Provider.Kayenta.CanaryConfigName {
+			canaryConfigId = s.Id
+			break
+		}
+	}
+
+	if len(canaryConfigId) == 0 {
+		err = errors.New("no matching canary config id")
```

If we return the canaryConfigId within the for loop, then we can assume that reaching this code path means that the canary config id was not found. As a result, we do not need this if statement as we can just return the err.

---

No.   that will not work if there is no match with the if.   Then there will be no return.